### DISCUSSION TITLE
Upload files

### DISCUSSION THREAD

**Discussion by rmfjd78** [h2oai/wave#583] [Q&A]
I'm new using the H2O Wave and have some questions regarding ui.file_upload:
1- Can we specify the destination folder? if not where are the files uploaded (folder)?
2- I've implemented a simple APP that have an option to upload files and after drag and drop the file the page passes to loading and doesn't stop.

**This is the code of my app**
`import os
from h2o_wave import Q, app, ui, main

sample_markdown = '''
XXXXXXX- Dataknoeldge Application

Esta aplicação serve de apoio á análise dados'''

example_dir = os.path.dirname(os.path.realpath(__file__))
_app_host = '127.0.0.1'
_app_port = '10102'
global localseparador

async def showpage(q: Q):
	q.page['meta'] = ui.meta_card(box='', layouts=[
		ui.layout(
			breakpoint='m',
			zones=[
				ui.zone('header', size='80px'),
				ui.zone('body', direction=ui.ZoneDirection.ROW, zones=[
					ui.zone('content'),
				], size='700px'),
				ui.zone('footer', size='80px'),
			]
		)
	])

	q.page['header'] = ui.header_card(
		# Place card in the header zone, regardless of viewport size.
		box='header',
		title='UTopIA 2.0 - Dataknowledge',
		subtitle='Ferramenta de análise de dados',
		nav=[
			ui.nav_group('Menu', items=[
				ui.nav_item(name='#menu/Home', label='Página Inicial'),
				ui.nav_item(name='#menu/Importacao', label='Importação ficheiro'),
				ui.nav_item(name='#menu/Correlacao', label='Mapa correlação'),
			]),
			ui.nav_group('Help', items=[
				ui.nav_item(name='#about', label='About', icon='Info'),
				ui.nav_item(name='#support', label='Support', icon='Help'),
			])
		],
	)

	q.page['Home'] = ui.markdown_card(
		box=ui.boxes('content', 'sidebar', 'sidebar'),
		title='',
		content=sample_markdown,
	)

	q.page['footer'] = ui.footer_card(box='footer', caption='(c) 2021 COFIDIS. All rights reserved.')

	await q.page.save()

async def show_example(q: Q):
	if q.args.submit:
		del q.page['Home']

		q.page['Home'] = ui.markdown_card(
			box=ui.boxes('content'),
			title='',
			content=sample_markdown,
		)
		localseparador = q.args.separador

	hash = q.args['#']
	if hash:
		if hash == 'menu/Home':
			del q.page['Home']

			q.page['Home'] = ui.markdown_card(
				box=ui.boxes('content'),
				title='',
				content=sample_markdown,
			)
		if hash == 'menu/Importacao':
			del q.page['Home']

			q.page['Home'] = ui.form_card(box=ui.box('content'),
										  items=[ui.text_xl(content='Importação de ficheiro'),
												 ui.combobox(name='separador',
															 label='Separador: ',
															 choices=[',', ';', ':'],
															 tooltip='Separador do ficheiro'),
												 ui.file_upload(name='user_files', label='Upload', multiple=False,
																file_extensions=['csv'], visible=True),
										         ui.button(name='submit', label='Submit', primary=True),
												 ])
		if hash == 'menu/Correlacao':
			del q.page['Home']

			q.page['Home'] = ui.markdown_card(
				box=ui.boxes('content'),
				title='',
				content='teste-' + str(q.args.separador),
			)
	await q.page.save()

@app('/Dataknowledge')
async def serve(q: Q):
	if not q.client.initialized:
		q.client.initialized = True
		await showpage(q)

	route = q.args['#']
	if not route:
		route = 'Home'

	await show_example(q)

print('----------------------------------------')
print(' Welcome to DataKnowledge')
print('')
print(' Go to http://localhost:10101/Dataknowledge')
print('----------------------------------------')
`

**Comment by mtanco** [comment]
Hello! Welcome to using Wave! Your app is looking nice so far, hope this helps: 

1. Where does uploaded data go?
   * When you upload a file in an app, it is added to the Wave Server, you can then download it from the server to anywhere you'd like in your development or runtime environment. 
   * This is done with the q.site.download `await q.site.download(url=q.args.user_files[0], path=./my_relative_path )`, where `user_files` is the `name` of the `ui.file_upload` object
   * You can then do whatever you wold like with that file

2. Why is my app stuck in `loading` after uploading data?
   * The File Upload widget comes with an upload button, so you do not have to have your own button, submit. When you click the Upload button an argument is sent to the server with the name of the upload object: `q.args.user_files`. You'll want to add a statement looking for this argument and then doing something with the data


Here is an app which allows a user to upload a CSV and then displays that data as a table, in case it helps to see an example: https://gist.github.com/mtanco/9c39438660eef8326f087d3a22e7e23b

Please let us know if you have any other questions!

**Comment by Rmfjd1978** [comment]
Can i cancel an upload while processing?

**Reply by azim-b** [reply]
this is answered here: https://github.com/h2oai/wave/discussions/1201

in short, no - this is currently not possible
